exercises/linked-list/.meta/description.md

```diff
@@ -0,0 +1,36 @@
+Implement a doubly linked list.
+
+Like an array, a linked list is a simple linear data structure. Several
+common data types can be implemented using linked lists, like queues,
+stacks, and associative arrays.
+
+A linked list is a collection of data elements called *nodes*. In a
+*singly linked list* each node holds a value and a link to the next node.
+In a *doubly linked list* each node also holds a link to the previous
+node.
+
+You will write an implementation of a doubly linked list. Implement a
+`Node` to hold a value and pointers to the next and previous nodes. Then
+implement a `List` which holds references to the first and last node and
+offers functions for adding and removing items.
+
+Your `Node` should have the following fields:
+
+* `Val`: the node's value (we will use `interface{}`).
+* `Next`: the next node.
+* `Prev`: the previous node.
+
+Your `List` should have the following fields and functions:
+
+* `Head`: the first element in the linked list.
+* `Tail`: the last element in the linked list.
+* `PushBack(v interface{})`: *insert value at back*.
+* `PopBack() (interface{}, error)`: *remove value at back*.
+* `PushFront(v interface{}) `: *remove value at front*.
+* `PopFront() (interface{}, error)`: *insert value at front*.
+* `Reverse()`: reverse the linked list (Head, Tail = Tail, Head).
+
+Instead of not covering error conditions, like calling `PopBack` or `PopFront` on an empty list,
+we will follow Go's idiomatic style and implement the error checks as well.
```

Comment:
@leenipper: I have overwritten the default `description.md`, so that the Go implementation is more idiomatic.

`Reverse()` was added into the requirement, as I feel it's where the we touch the  "doubly" aspect of the implementation.

Reply:
> significantly touch the "doubly" aspect.

Agree.

An on that "doubly" aspect, if we have First() as suggested, then perhaps also have a Last(), along with a Previous() ?  This assumes exposing less of the struct members for protection of package implemenation.  Otherwise, user can manually traverse. I like exposing less internals.

Reply:
>  This assumes exposing less of the struct members for protection of package implemenation. Otherwise, user can manually traverse. I like exposing less internals.

Agree. I was confused because we still return a `*Node` (somehow I associated this with arbitrary content manipulation). It's not the case and package user should not be able to manipulate a node's `next` and `prev`.

So I will change it to:

```go
type Node struct {
    Val interface{}
    prev *Node
    next *Node
}

func (n *Node) First() *Node { ... } // traverse till head
func (n *Node) Last() *Node { ... } // traverse till tail
func (n *Node) Next() *Node { ... }
func (n *Node) Prev() *Node { ... }
```

I will also change the linked list definition to use getter functions, so we are consistent.

```go
type List struct {
    head *Node
    tail *Node
}

func (ll *List) Head() *Node { ... }
func (ll *List) Tail() *Node { ... }
```